File: src/services/id-generator.ts
import * as path from 'node:path';
import { readJson, writeJson, exists } from './storage.js';
import { getAdaptDir, ensureDir } from '../lib/utils.js';

type IdPrefix = 'obs' | 'ana' | 'adp' | 'plan';

interface CounterState {
  obs: number;
  ana: number;
  adp: number;
  plan: number;
}

function getCounterPath(): string {
  return path.join(getAdaptDir(), 'state', 'counter.json');
}

function loadCounter(): CounterState {
  const counterPath = getCounterPath();
  if (!exists(counterPath)) {
    return { obs: 0, ana: 0, adp: 0, plan: 0 };
  }
  return readJson<CounterState>(counterPath);
}

function saveCounter(counter: CounterState): void {
  const counterPath = getCounterPath();
  ensureDir(path.dirname(counterPath));
  writeJson(counterPath, counter);
}

/**
 * Generate a unique ID with the given prefix.
 *
 * Format: `{prefix}_{year}_{sequence}` where sequence is zero-padded to 3 digits.
 * Example: `obs_2026_001`
 *
 * Reads the current counter from `.adapt/state/counter.json`, increments it,
 * writes it back, and returns the formatted ID.
 */
export function generateId(prefix: IdPrefix): string {
  const counter = loadCounter();
  counter[prefix] += 1;
  saveCounter(counter);

  const year = new Date().getFullYear();
  const sequence = String(counter[prefix]).padStart(3, '0');

  return `${prefix}_${year}_${sequence}`;
}

File: code_adapt/services/id_generator.py
"""Sequential ID generation: obs_2026_001 format."""

from __future__ import annotations

from datetime import datetime
from pathlib import Path
from typing import Literal

from ..storage import get_adapt_dir, ensure_dir, read_json, write_json

IdPrefix = Literal["obs", "ana", "adp", "plan"]


def _counter_path() -> Path:
    return get_adapt_dir() / "state" / "counter.json"


def _load_counter() -> dict[str, int]:
    path = _counter_path()
    if not path.exists():
        return {"obs": 0, "ana": 0, "adp": 0, "plan": 0}
    return read_json(path)


def _save_counter(counter: dict[str, int]) -> None:
    path = _counter_path()
    ensure_dir(path.parent)
    write_json(path, counter)


def generate_id(prefix: IdPrefix) -> str:
    counter = _load_counter()
    counter[prefix] = counter.get(prefix, 0) + 1
    _save_counter(counter)
    year = datetime.now().year
    seq = str(counter[prefix]).zfill(3)
    return f"{prefix}_{year}_{seq}"
